From 781e838721c1cf66f95393b4f6ec8030589af6a0 Mon Sep 17 00:00:00 2001
From: Brian Silverman <bsilverman@iveia.com>
Date: Tue, 7 Apr 2020 13:55:15 -0400
Subject: [PATCH] zynqmp: blank out repeated boot_targets

The env arg boot_targets is prepended with the current booting device.  As
such, it tries to boot twice from that device if it is already in the
boot_targets list.  This fix blanks out any repeats of the prepended device.
---
 board/xilinx/zynqmp/zynqmp.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/board/xilinx/zynqmp/zynqmp.c b/board/xilinx/zynqmp/zynqmp.c
index 853f4f9..c46a162 100644
--- a/board/xilinx/zynqmp/zynqmp.c
+++ b/board/xilinx/zynqmp/zynqmp.c
@@ -701,6 +701,25 @@ int board_late_init(void)
 		sprintf(new_targets, "%s %s", mode,
 			env_targets ? env_targets : "");
 
+    /*
+     * If the added first target is already in boot_targets, blank out any
+     * repeated targets
+     */
+    int i;
+    for (i = 0; new_targets[i]; i++) {
+        if (new_targets[i] == ' ') break;
+    }
+    int first_target_len = i;
+    int new_targets_len = strlen(new_targets);
+    for (int i = 1; i < new_targets_len - first_target_len + 1; i++) {
+        int match = strncmp(&new_targets[0], &new_targets[i], first_target_len) == 0;
+        int terminator = new_targets[i+first_target_len] == '\0'
+            || new_targets[i+first_target_len] == ' ';
+        if (match && terminator) {
+            memset(&new_targets[i], ' ', first_target_len);
+        }
+    }
+
 	env_set("boot_targets", new_targets);
 
 	initrd_hi = gd->start_addr_sp - CONFIG_STACK_SIZE;
-- 
2.7.4

